### ISSUE TITLE
Feature request: require custom elements be registered on HTMLElementTagNameMap

### ISSUE THREAD

**Issue by rictic** [runem/lit-analyzer#73]
A pass that would diagnose when a custom element is recognized in source code but it hasn't be declared on the HTMLElementTagNameMap, or if it is declared with the wrong tag name. For example:

Warns:
```typescript
import {customElement, LitElement} from 'lit-element';
@customElement('foo-bar')
class FooBar extends LitElement {}
//    ~~~~~~  warning!
```

Warns:
```typescript
import {customElement, LitElement} from 'lit-element';
@customElement('foo-bar')
class FooBar extends LitElement {}

declare global {
  interface HTMLElementTagNameMap {
    'fizz-buzz': FooBar;
//  ~~~~~~~~~~~ warning!
  }
}
```

Does not warn:
```typescript
import {customElement, LitElement} from 'lit-element';
@customElement('foo-bar')
class FooBar extends LitElement {}

declare global {
  interface HTMLElementTagNameMap {
    'foo-bar': FooBar;
  }
}
```

**Comment by runem** [comment]
Great idea, I love this rule! This would be possible to implement after the [performance](https://github.com/runem/lit-analyzer/tree/performance) branch has been finished which includes the new version of `web-component-analyzer`. The new version of WCA emits all definitions of the same element and not just the last one found.

**Comment by e111077** [comment]
Alternative, what's wrong with adding types to `@fires`?

```ts
// my-element.ts
/**
 * @fires my-event {MyEventDetail}
 */
export class MyEvent extends LitElement { ... }
```

```ts
// main.ts
html`
  <!-- type checking happens here -->
  <my-element @my-event=${this.onMyEvent}></my-element>
...
`;

onMyEvent(e: MyEventDetail) { ... }
```

**Comment by rictic** [comment]
+1 to checking events (I believe there's an event map as well, though it's not element-specific), but that should be a different issue IMO

**Comment by e111077** [comment]
will file

**Comment by runem** [comment]
I'm currently working on this branch: https://github.com/runem/lit-analyzer/tree/1.2.0 which should be able to support building this rule :-)

**Comment by e111077** [comment]
oh wow I'm dumb; I totally read this as Custom Events not CustomElements haha >.< issue filed on #85

**Comment by runem** [comment]
I just implemented the rule on [1.2.0](https://github.com/runem/lit-analyzer/pull/112) :-)

The rule is called `no-missing-element-type-definition` and is disabled by default both when running strict and non-strict. Please let me know if you have any comments or any additions that you want me to add to this feature.

![demo](https://user-images.githubusercontent.com/5372940/86847333-68993380-c0ac-11ea-8325-61137342fb25.gif)

**Comment by justinfagnani** [comment]
That quickfix is awesome!!

**Comment by rictic** [comment]
Awesome!

**Comment by WickyNilliams** [comment]
I've not been able to get this working - is it broken with `lit@2`. See the attached screenshot. No error reported

![image](https://user-images.githubusercontent.com/1091390/143035676-21bbb65d-78aa-4210-bb0b-184f8824264a.png)

**Comment by WickyNilliams** [comment]
Sorry I take it back, it is working now. Not sure what happened there!
